docs/rfc/0698-cli-infrastructure.adoc

```diff
@@ -0,0 +1,297 @@
+
+. One binary that defines the core subcommands
+. Many binaries that take advantage of the subcommand mechanism
+
+Pros of *1.*:
+
+* One-stop shop for all core functionality
+* One binary is packaged and distributed
+
+Cons of *1.*:
+
+* Making changes to one subset of the commands could mean a version
+  change of the whole binary
+
+Pros of *2.*:
+
+* We will be organising commands as libraries (as per RFC-682) — so
+  modular commands would mirror this structure well.
+* Subcommands can be versioned independently of each other.
+* We can dogfood our subcommand functionality
+
+Cons of *2.*:
+
+* Makes it easier to override one of the core commands.
+* Many binaries must be packaged and distributed.
+
+The author recommends *1.*. As mentioned, we have a hard requirement
+of not allowing the core commands to be overridden. In *2.* this is
+_always_ possible.
+
+=== `radicle-bins`
+
+Per <<arch-cli, RFC-682#CLI>>, all subcommands will be provided as
+libraries. However, this leaves the question of where the binary might
+live. So far, it has been an unspoken formalism that all binaries live
+in `radicle-bins` where we can provide a `Cargo.lock` file. All
+functionality will live in `radicle-link`, including the `main`
+function, and the entry in `radicle-bins` will be a thin wrapper
+around this function.
+
+== Core
+
+[quote]
+------
+We swears to serve the master of the precious. We will swear on
+the…on the precious.
+------
+-- Gollum, The Two Towers
+
+With all the details in place, we define a non-exhaustive list of the
+core commands that we wish to include for the first iteration. More
+components MAY be added as the result of future RFCs. We SHALL keep a
+list of the core commands as part of the `radicle-link` documentation.
+
+=== `rad-identities`
+
+The `rad identities` subcommand provides the management of identity
+documents and their metadata. It is further split into subcommands
+which we sketch out below:
+
+* `project [ create | get | list | update | checkout | review ]` --
+  modify and retrieve `Project` identities.
+* `person [ create | get | list | update | checkout | review ]` --
+  modify and retrieve `Person` identities.
+* `any [ get | list ]` -- retrieve identities where the caller does not
+  know what kind of identity it is. The CLI will make a best-effort to
+  infer the type, or else return an opaque blob.
+* `local [ set | get | default ]` -- modify or retrieve a local identity
+* `tracking [ track | untrack | list ]` -- modify and list the tracking
+  graph for a given identity.
+* `rad-refs [ self | signed | delegates | delegate ]` -- retrieve the
+  blobs that live under `rad/<refname>`
+* `refs [ heads | notes | tags | category ]` -- list the reference names.
+  under the respective category, where `category` is a user supplied category.
+
+=== `rad-replicate`
+
+* `ask` -- tell the `daemon` that you wish to replicate the
+  given URN. Pending <<rl-141, radicle-link#141>>.
```

Comment:
To implement a list of identities we want to replicate but haven’t found yet we would need a corresponding command. Something like `rad replicate status`.

Reply:
Great point :+1: